tempesta_fw/http_sticky.c

```diff
@@ -0,0 +1,740 @@
+			memcpy(page_address(frag->page.p) + page_offset,
+			       data + data_offset, copy_size);
+			skb_frag_size_add(frag, copy_size);
+			data_offset += copy_size;
+			len -= copy_size;
+		}
+		/*
+		 * The above for() loop runs at least once,
+		 * which means that i_frags is always incremented.
+		 */
+		state->fragnum = i_frag - 1;
+	}
+	/* In the end, data_offset equals the initial len value */
+	skb->len += data_offset;
+	skb->data_len += data_offset;
+}
+
+static int
+tfw_http_msg_setup(tfw_mastate_t *state, TfwMsg *msg, size_t len)
+{
+	struct page *page;
+	struct sk_buff *skb;
+	int i_frag, i_skb, nr_skb_frags;
+	int nr_frags = DIV_ROUND_UP(len, PAGE_SIZE);
+	int nr_skbs = DIV_ROUND_UP(nr_frags, MAX_SKB_FRAGS);
+
+	/*
+	 * TODO: Make sure to create SKBs with payload size <= MSS
+	 */
+	for (i_skb = 0; i_skb < nr_skbs; i_skb++) {
+		if ((skb = alloc_skb(MAX_TCP_HEADER, GFP_ATOMIC)) == NULL) {
+			return -ENOMEM;
+		}
+		skb_reserve(skb, MAX_TCP_HEADER);
+		ss_skb_queue_tail(&msg->skb_list, skb);
+
+		nr_skb_frags = min_t(size_t, nr_frags, MAX_SKB_FRAGS);
+		for (i_frag = 0; i_frag < nr_skb_frags; i_frag++) {
+			if ((page = alloc_page(GFP_ATOMIC)) == NULL) {
+				return -ENOMEM;
+			}
+			get_page(page);
+			skb_fill_page_desc(skb, i_frag, page, 0, 0);
+			skb->truesize += PAGE_SIZE;
+			skb_shinfo(skb)->nr_frags++;
+		}
+		nr_frags -= nr_skb_frags;
+	}
+	/* Set up initial state */
+	state->skb = ss_skb_peek(&msg->skb_list);
+	state->fragnum = 0;
+
+	return 0;
+}
+
+#define S_CRLF		"\r\n"
+#define S_CRLFCRLF	"\r\n\r\n"
+#define S_HTTP		"http://"
+
+#define S_302		"HTTP/1.1 302 Found"
+#define S_502		"HTTP/1.1 502 Bad Gateway"
+
+#define S_F_HOST		"Host: "
+#define S_F_DATE		"Date: "
+#define S_F_CONTENT_LENGTH	"Content-Length: "
+#define S_F_LOCATION		"Location: "
+#define S_F_CONNECTION		"Connection: "
+#define S_F_SET_COOKIE		"Set-Cookie: "
+
+#define S_V_DATE		"Sun, 06 Nov 1994 08:49:37 GMT"
+#define S_V_CONTENT_LENGTH	"9999"
+
+#define SLEN(s)		(sizeof(s) - 1)
+
+#define	S_302_FIXLEN							\
+	SLEN(S_302)							\
+	+ SLEN(S_CRLF)							\
+	+ SLEN(S_F_DATE) + SLEN(S_V_DATE)				\
+	+ SLEN(S_CRLF)							\
+	+ SLEN(S_F_CONTENT_LENGTH) + SLEN(S_V_CONTENT_LENGTH)		\
+	+ SLEN(S_CRLF)							\
+	+ SLEN(S_F_LOCATION) + SLEN(S_HTTP)				\
+	+ SLEN(S_CRLF)							\
+	+ SLEN(S_F_SET_COOKIE)						\
+	+ SLEN(S_CRLFCRLF)
+
+static size_t
+tfw_http_prep_date(u_char *buf)
+{
+	/* TODO: put in the current date stamp */
+	memcpy(buf, S_V_DATE, SLEN(S_V_DATE));
+	return SLEN(S_V_DATE);
+}
+
+static size_t
+tfw_http_prep_sticky(u_char *buf, u_char *value, size_t len)
+{
+	u_char *ptr = buf;
+
+	while (len--) {
+		snprintf(ptr, 100, "%02x", (*value++ & 0x0FF));
+		ptr += 2;
+	}
+	return (ptr - buf);
+}
+
+/*
+ * Send an HTTP 302 response to the client. The response redirects
+ * the client to the same URI as the original request, but includes
+ * 'Set-Cookie:' header field that sets Tempesta sticky cookie.
+ */
+static int
+tfw_http_send_302(TfwHttpMsg *hm)
```

Comment:
This should be more generic function which creates particular HTTP error response depending on passed arguments. Also should be moved to http.c